src/file_pipe_log/pipe.rs

```diff
@@ -8,23 +8,75 @@ use std::sync::Arc;
 use crossbeam::utils::CachePadded;
 use fail::fail_point;
 use log::error;
-use num_traits::FromPrimitive;
 use parking_lot::{Mutex, MutexGuard, RwLock};
 
 use crate::config::Config;
 use crate::env::FileSystem;
 use crate::event_listener::EventListener;
 use crate::metrics::*;
-use crate::pipe_log::{FileBlockHandle, FileId, FileSeq, LogQueue, PipeLog};
+use crate::pipe_log::{
+    FileBlockHandle, FileId, FileSeq, LogFileContext, LogQueue, PipeLog, Version,
+};
 use crate::{perf_context, Error, Result};
 
-use super::format::{FileNameExt, Version};
+use super::format::FileNameExt;
 use super::log_file::{build_file_reader, build_file_writer, FileHandler, LogFileWriter};
 
 struct FileCollection<F: FileSystem> {
+    /// Sequence number of the first file.
     first_seq: FileSeq,
-    active_seq: FileSeq,
+    /// Sequence number of the first file that is in use.
+    first_seq_in_use: FileSeq,
     fds: VecDeque<FileHandler<F>>,
+    /// `0` => no capbility for recycling stale files
+    /// `_` => finite volume for recycling stale files
+    capacity: usize,
+}
+
+#[cfg(test)]
+impl<F: FileSystem> Default for FileCollection<F> {
+    fn default() -> Self {
+        Self {
+            first_seq: 0,
+            first_seq_in_use: 0,
+            fds: VecDeque::new(),
+            capacity: 0,
+        }
+    }
+}
+
+impl<F: FileSystem> FileCollection<F> {
+    /// Recycle the first obsolete(stale) file and renewed with new FileId.
+    ///
+    /// Attention please, the recycled file would be automatically `renamed` in
+    /// this func.
+    pub fn recycle_one_file(&mut self, file_system: &F, dir_path: &str, dst_fd: FileId) -> bool {
```

Comment:
You must first "force_reset" the file before renaming. Otherwise the obsolete file will appear as the newest file and break recovery (this case should be tested, i.e. `self.file_system.open(&new_path)` fails and engine restarts).

Reply:
I didn't get it. In my opinion, the way, first with `rename` and followed with `force_reset` is better, which can tolerate this case.
* `force_reset` is just for rewritting the `header` with new format, such as V1 -> V1, V1 -> V2 or V2-> V1.
* About the priorities of `rename` and `force_reset`:
*  1. In the current design, `rename` the file first and `force_reset` next. The circumstance that the obsolete file is renamed with newest one won't break recovery if we set the mode with `tolerate_tail_corruption`.
* 2. In contrast to the original desgin, given that we `force_reset` firstly, and next to `rename`. If the `force_reset` is finished but `rename` failed, the obsolete file will have a rewritten header, i.e. converted from V1 -> V2, and unfortunately, the engine crashed( The obsolete file was not be cleaned by `Drop`. ). Then, we wanna restart the engine. Boom! It crashed again with the err, "The head file cannot be correctly parsed and `logbatch`s in it cannot be decoded with V2( because they are written with V1. )".

Reply:
> The circumstance that the obsolete file is renamed with newest one won't break recovery if we set the mode with tolerate_tail_corruption.

This is incorrect. recovery_mode only tolerates physical corruption. But the case I proposed will fail in logical verification (panic in memtable).

You are right that `rename` failure is not considered. How about renaming it into a tmp file and force_reset and rename it into new name? In fact, you can do the first rename early when the file is retired.

Reply:
> This is incorrect. recovery_mode only tolerates physical corruption. But the case I proposed will fail in logical verification (panic in memtable).
Assume that we still use the original design here, it can tolerate the recovery failure, triggered by recovering obsolete records. Because, (I.E. we have one recycled file from seq: 1 -> seq: 20, and after `rename`, the Engine crashed. )
* Only `Version::V2` supports log files to be recycled.
* The records in the newest recycled log file, which haven't been rewritten yet, only has signatures generated by the oldest `file_seq`, that is `seq: 1`. And these signatures would be verified with the newest `seq_num: 20` and returns with errs (err: `Checksum expected ... but got ... `) while recovering. And the caller, `reader.next()`, will get the err, and make the judgement that it can tolerate it if the recovery_mode == `TolerateTailCorruption`.

So, the original design is good enough to meet the requirements for tackling these unexpected failures.

Reply:
> Only Version::V2 supports log files to be recycled.

This is incorrect. We only guarantee the new file (after being force_reset) is V2, but the old file doesn't have to be V2. So there's no signature to protect the logs before being force_reset.

But you remind me that this case doesn't have to be handled my way. You can simply add a logic to purge files in range [first_seq, seq2] if seq2 is V1.

Reply:
I got your point.
What u emphasised above was the case that one stale file was recycled to V2 (config.format_version was changed before restarting the Engine.), but the `force_reset` failed and the Engine crashed. So, the newest file was truely renamed but still with format_version v1. And the recovery truely failed with panic in memtable recovery.

> You can simply add a logic to purge files in range [first_seq, seq2] if seq2 is V1.
Your suggestion is that we just directly remove these obsolete V1 files ?

Reply:
Yes.